### PR TITLE
Simplified Scaffolding based on the builder pattern

### PR DESCRIPTION
**Pull request by droot**
This PR does the following:
 
- move existing scaffolding under pkg `pkg/scaffold/v1`
- introduces `v2` pkg for builder pattern based scaffolding
- Refactors scaffolding logic and move it from `cmd` to `pkg`

Note: There is more work to be done to make builder pattern based scaffolding for ex. controller-gen tools doesn't work with new path structure, scaffolding-test for `v2` aren't working yet.

### CODE REVIEW

File: test/project/PROJECT [hunk 1]
```diff
@@ -1,3 +1,4 @@
 version: "1"
```
Comment (mengqiy) [inline test/project/PROJECT]: What is this `version` used for? Have we used it somewhere?

Comment (droot) [reply test/project/PROJECT]: We used this to differentiate between the `pre-1.0` and `1.0` projects. We can probably use this field as `scaffoldingVersion` as well, I don't see any downside.

Comment (mengqiy) [reply test/project/PROJECT]: Yes, I guess we can use `version` field for `scaffoldingVersion` purpose.

**Comment by mengqiy** [comment]
LGTM
It seems some tests are not passing any more.

**Comment by droot** [comment]
@mengqiy I have updated the PR to include the following: 

- move existing scaffolding under pkg pkg/scaffold/v1
- introduces v2 pkg for builder pattern based scaffolding
- Refactors scaffolding logic and move it from cmd to pkg

Note: There is more work to be done to make builder pattern based scaffolding for ex. controller-gen tools doesn't work with new path structure, scaffolding-test for v2 aren't working yet. But I would like to keep the scope for this PR to current set of changes. The new changes be build on top of this.

**Comment by mengqiy** [comment]
It seems the tests are still failing with legitimate reason.
Are we expect more changes on this PR?

File: test/project/PROJECT (hunk 1 above)
Comment (droot) [reply test/project/PROJECT]: Done.

**Comment by droot** [comment]
@mengqiy Added second commit which implements scaffolding changes needed for v2.

File: test/project_v2/api/v1/doc.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package v1 contains API Schema definitions for the crew v1 API group
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen=package,register
+// +k8s:conversion-gen=sigs.k8s.io/kubebuilder/test/project_v2/pkg/apis/crew
+// +k8s:defaulter-gen=TypeMeta
+// +groupName=crew.testproject.org
```
Comment (mengqiy) [inline test/project_v2/api/v1/doc.go]: https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/simplified-scaffolding.md is designed for handling one Group, but it seem we have multiple groups under `/api`.

File: test/project_v2/main.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	crewv1 "sigs.k8s.io/kubebuilder/test/project_v2/api/v1"
+	policyv1beta1 "sigs.k8s.io/kubebuilder/test/project_v2/api/v1beta1"
+	shipv1beta1 "sigs.k8s.io/kubebuilder/test/project_v2/api/v1beta1"
+	creaturesv2alpha1 "sigs.k8s.io/kubebuilder/test/project_v2/api/v2alpha1"
+	"sigs.k8s.io/kubebuilder/test/project_v2/controllers"
+	// +kubebuilder:scaffold:imports
```
[the remark this answers is not in the document]
Comment (droot) [reply test/project_v2/main.go]: This is a marker comment used for adding import statement when we add controller for a new API. This is not processed by the controller-tools or anything. I thought of adding some comments just above this marker, but that is going to complicate the dynamic insertion of code fragments on addition of new controllers, so left it like that.

File: test.sh [hunk 4]
```diff
@@ -99,40 +96,40 @@ setup_envs
...
 test_create_namespaced_coretype_controller
 
 cd ${go_workspace}/src/sigs.k8s.io/kubebuilder
 
 go test ./cmd/... ./pkg/...
 
-cd test/project
-make
-cd -
+./generated_golden.sh
```
[the remark this answers is not in the document]
Comment (droot) [reply test.sh]: Actually, this needs to be taken out. Instead, I should be exploding the vendor in each test project and running make. Thanks for pointing it out.

File: test/project_v2/main.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+)
+
+func init() {
+
+	crewv1.AddToScheme(scheme)
+	shipv1beta1.AddToScheme(scheme)
+	creaturesv2alpha1.AddToScheme(scheme)
+	corev1.AddToScheme(scheme)
+	policyv1beta1.AddToScheme(scheme)
+	// +kubebuilder:scaffold:scheme
```
[the remark this answers is not in the document]
Comment (droot) [reply test/project_v2/main.go]: Same comment as for `+kubebuilder:scaffold:imports`

File: test/project_v2/api/v1/doc.go (hunk 2 above)
Comment (droot) [reply test/project_v2/api/v1/doc.go]: Aah, completely missed that aspect. Will take a look.

Comment (droot) [reply test/project_v2/api/v1/doc.go]: Just realized, haven't addressed this one, taking a look now.

File: test.sh [hunk 6]
```diff
@@ -21,17 +21,14 @@ source common.sh
 
 function test_init_project {
   header_text "performing init project"
-  kubebuilder init --domain example.com <<< "y"
-  make
-  cache_dep
+  kubebuilder init --domain example.com <<< "n"
+  # make
```
Comment (mengqiy) [inline test.sh]: drop the commented line

File: test.sh [hunk 7]
```diff
@@ -21,17 +21,14 @@ source common.sh
...
 
-
-
-function test_init_project_manual_dep_ensure {
-  header_text "performing init project w/o dep ensure"
-  kubebuilder init --domain example.com <<< "n"
-  dep ensure -v
+function test_make_project {
+  header_text "running make in project"
+  # kubebuilder init --domain example.com <<< "n"
```
Comment (mengqiy) [inline test.sh]: drop the commented lines

**Comment by mengqiy** [comment]
Please squash after address these 2 nits. Good to go.

**Pull request merged**